fix: Build a new workshop for debugging using throw/try/catch/finally in js (#67750)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/blocks/workshop-library-search/6a1d2f52249f2e98cceff74b.md

---
id: 6a1d2f52249f2e98cceff74b
title: Step 13
challengeType: 0
dashedName: step-13
---

# --description--

Awesome! Now it is time for a final polish. You need to clean up your debugging tools and prepare to output a final status message.

First, delete the `debugger;` statement from your `catch` block.

Next, you may notice that there is a paragraph element below the search bar in your HTML. It has an `id` of `scanStatus`. 

> Before writing the final logic to update it, you need to grab this new element. At the top of your code, underneath your other DOM variables, declare a variable named `scanStatusDisplay` and assign it the element `scanStatus` using `document.getElementById()`.

Reviewer: the element described here does not exist, it appears in the next step tho

Author: added the scanStatus element to the HTML in all files in the latest commit

---
id: 6a1d2f52249f2e98cceff74b
title: Step 10
challengeType: 0
dashedName: step-10
---

# --description--

Awesome! Now it is time for a final polish. You need to clean up your debugging tools and prepare to output a final status message.

First, delete the `debugger;` statement from your `catch` block.

Next, you may notice that there is a paragraph element below the search bar in your HTML. It has an `id` of `scanStatus`. 

> Before writing the final logic to update it, you need to grab this new element. At the top of your code, underneath your other DOM variables, declare a variable named `scanStatusDisplay` and assign it the element `scanStatus` using `document.getElementById()`.

Reviewer 2: "grab" is more informal than the rest of the workshop ("you need to handle", "you need to wrap"), and "assign it the element `scanStatus`" reads as though the element itself is named `scanStatus` rather than carrying that `id`. Line 14 also has trailing whitespace.

```suggestion
Next, you may notice that there is a paragraph element below the search bar in your HTML. It has an `id` of `scanStatus`.

Before writing the final logic to update it, you need to select this new element. At the top of your code, underneath your other DOM variables, declare a variable named `scanStatusDisplay` and assign it the element with the `scanStatus` id using `document.getElementById()`.
```

# --hints--

You should remove the `debugger;` statement from your code.

```js

> assert.isFalse(/debugger/.test(code));

Reviewer 2: `assert.notMatch` is the semantic form here and produces a usable failure message, where `isFalse` just reports that `false` was expected:

```suggestion
assert.notMatch(code, /debugger/);
```

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Library Book Search</title>
    <link rel="stylesheet" href="styles.css" />
  </head>
  <body>
    <div>
      <h1>Find book title</h1>
      <input type="text" id="bookTitleInput" placeholder="Enter book title">
      <button id="search">Search</button>
      <p id="statusMessage"></p>
      <p id="scanStatus"></p>
    </div>
    <script src="script.js"></script>
  </body>
</html>
```

```css
body {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
}

input {
  padding: 0.5rem;
  width: 200px;
}

button {
  padding: 0.5rem 1rem;
  cursor: pointer;
  background-color: #000000;
  color: white;
}

#statusMessage {
  margin-top: 1.5rem;
  font-weight: bold;
  height: 1.5rem;
}
```

```js
const library = [
  "The Hobbit", "1984", "Dune", "Fahrenheit 451", "To Kill a Mockingbird",
  "The Great Gatsby", "Moby Dick", "Pride and Prejudice", "Frankenstein"
];

const searchBtn = document.getElementById('search');
const statusDisplay = document.getElementById('statusMessage');

> const bookInput = document.getElementById('bookTitleInput');

Reviewer: ```suggestion
const searchBtn = document.getElementById("search");
const statusDisplay = document.getElementById("statusMessage");
const bookInput = document.getElementById("bookTitleInput");
```

--fcc-editable-region--

const findBook = (title) => {
  if (title === "") {
    throw new Error("Search field cannot be empty");
  }

  if (!library.includes(title)) {
    throw new Error(title + " not found!");
  }

  return title + " is available to borrow!";
}

> searchBtn.addEventListener('click', () => {

Reviewer: ```suggestion
};

searchBtn.addEventListener("click", () => {
```

  try {
    const result = findBook(bookInput.value.trim());
    statusDisplay.textContent = result;
  } catch (err) {
    statusDisplay.textContent = err.message;
    console.error(err);
    debugger;
--fcc-editable-region--
  } finally {
    bookInput.value = "";

>   }

Reviewer: ```suggestion
    bookInput.value = "";
  }
```
